Generate serializers for DDScannerResult and NSValue

beidson opened the pull request:
#### 72c2238b339fd7c807bb5f0d7f5534cb9cfc8f4f
<pre>
Generate serializers for DDScannerResult and NSValue
<a href="https://bugs.webkit.org/show_bug.cgi?id=264915">https://bugs.webkit.org/show_bug.cgi?id=264915</a>
<a href="https://rdar.apple.com/118486039">rdar://118486039</a>

Reviewed by Alex Christensen.

This patch starts the process of breaking out specific secure coding types into their own wrappers.

Starting with DDScannerResult, it optionally uses SPI on the object if it exists to grab a directly
serializable NSDictionary, or it falls back to just wrapping the object itself like any secure coding type.

That NSDictionary is known to contain NSValues that wrap an NSRange, so this patch also adds an NSValue wrapper.

Currently, any NSValues that are serialized would go through the secure coding path.
For this one wrapped type we&apos;re known to need (NSRange), we go through a directly serialized code path.

Headed down this road, the goal is to identify all specific types we need to serialize, have a direct wrapper
for each, and reduce our reliance on the secure coding code path to zero.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.h:
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::getClass&lt;DDScannerResult&gt;):
(IPC::typeFromObject):
* Source/WebKit/Shared/Cocoa/CoreIPCDDScannerResult.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCDictionary.h.
(WebKit::CoreIPCDDScannerResult::CoreIPCDDScannerResult):
* Source/WebKit/Shared/Cocoa/CoreIPCDDScannerResult.mm: Copied from Source/WebKit/Shared/Cocoa/CoreIPCSecureCoding.mm.
(WebKit::shouldWrapDDScannerResult):
(WebKit::CoreIPCDDScannerResult::CoreIPCDDScannerResult):
(WebKit::CoreIPCDDScannerResult::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCDDScannerResult.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCDictionary.h:
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.h:
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.mm:
(WebKit::valueFromID):
* Source/WebKit/Shared/Cocoa/CoreIPCNSValue.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCDictionary.h.
(WebKit::CoreIPCNSValue::CoreIPCNSValue):
* Source/WebKit/Shared/Cocoa/CoreIPCNSValue.mm: Copied from Source/WebKit/Shared/Cocoa/CoreIPCSecureCoding.mm.
(WebKit::CoreIPCNSValue::CoreIPCNSValue):
(WebKit::CoreIPCNSValue::toID const):
(WebKit::CoreIPCNSValue::shouldWrapValue):
* Source/WebKit/Shared/Cocoa/CoreIPCNSValue.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCSecureCoding.h:
(WebKit::CoreIPCSecureCoding::CoreIPCSecureCoding):
* Source/WebKit/Shared/Cocoa/CoreIPCSecureCoding.mm:
(WebKit::CoreIPCSecureCoding::CoreIPCSecureCoding):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.mm:
(IPC::getClass&lt;DDScannerResult&gt;): Deleted.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Since NSNumber is itself an NSValue, refactor these tests to explicitly emplace into the testing variant:
* Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
(TEST):
(fakeDataDetectorResultForTesting):

Canonical link: <a href="https://commits.webkit.org/270840@main">https://commits.webkit.org/270840@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/acbd61923934cf9f85b6838189d2c35e125a6082

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26559 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5175 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27807 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28772 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24290 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/26969 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6983 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2582 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/24252 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26820 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4007 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22820 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3531 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3572 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/23806 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29254 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24220 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24204 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29840 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3604 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1807 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27734 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5040 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6377 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4078 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3932 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
```diff
@@ -270,6 +277,8 @@ NSType typeFromObject(id object)
         return NSType::Array;
     if ([object isKindOfClass:[WebCore::CocoaColor class]])
         return NSType::Color;
+    if ([object isKindOfClass:[PAL::getDDScannerResultClass() class]])
```
PAL::isDataDetectorsCoreFrameworkAvailable() &&

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSValue.serialization.in:
```diff
@@ -0,0 +1,37 @@
+}
+
+[WebKitPlatform] class WebKit::CoreIPCNSValue {
+    std::optional<WebKit::CoreIPCNSValue::WrappedValue> m_wrappedValue;
```
Shouldn't this be a std::variant<WebKit::CoreIPCNSValue::WrappedValue, WebKit::CoreIPCSecureCoding> since you will always only have one or the other?

beidson replied:
s/Shouldn't/Couldn't/

But yes, agreed, it can be!

achristensen07 replied:
If they were both null/nullopt then an assertion would fail and we would dereference null.  Using std::variant would eliminate this possibility.

beidson replied:
Yah, I made that change (while still tracking down TV/Watch build failures)

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCDDScannerResult.mm:
```diff
@@ -0,0 +1,76 @@
+#import "config.h"
+#import "CoreIPCDDScannerResult.h"
+
+#if ENABLE(DATA_DETECTION)
+#if PLATFORM(COCOA)
```
&&

beidson replied:
Yup. Ugh.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCDDScannerResult.mm:
```diff
@@ -0,0 +1,76 @@
+CoreIPCDDScannerResult::CoreIPCDDScannerResult(DDScannerResult *result)
+{
+    if (shouldWrapDDScannerResult())
+        m_plist = makeUnique<CoreIPCDictionary>([result _webKitPropertyListData]);
```
I think it would be wise to pave the way for adding a compile-time macro to make it so that this is the only way CoreIPC will accept the serialization on platforms that support it.

beidson replied:
It necessarily cannot be a compile time macro.
At least, in the mid-term future, it will need to be a runtime switch.

achristensen07 replied:
We are working to make it a compile time macro eventually, though, to remove the current decoding code.

beidson replied:
We can't make a compile-time macro now, anyways, since these checks are only possible at runtime.

achristensen07 commented on Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.h:
```diff
@@ -66,12 +73,16 @@ enum class NSType : uint8_t {
     SecureCoding,
     String,
     URL,
+    NSValue,
```
Do we need this if DATA_DETECTION is not enabled?

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCDDScannerResult.h:
```diff
@@ -0,0 +1,70 @@
+#pragma once
+
+#if ENABLE(DATA_DETECTION)
+#if PLATFORM(COCOA)
```
I'm not sure this is needed.  DATA_DETECTION is only enabled on a subset of Cocoa platforms.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCDDScannerResult.mm:
```diff
@@ -0,0 +1,87 @@
+#import <pal/cocoa/DataDetectorsCoreSoftLink.h>
+
+@interface DDScannerResult ()
+- (NSDictionary *)_webKitPropertyListData;
```
This should probably go into an SPI header at some point with availability HAVE macros

beidson committed:
Generate serializers for DDScannerResult and NSValue
https://bugs.webkit.org/show_bug.cgi?id=264915
rdar://118486039

Reviewed by Alex Christensen.

This patch starts the process of breaking out specific secure coding types into their own wrappers.

Starting with DDScannerResult, it optionally uses SPI on the object if it exists to grab a directly
serializable NSDictionary, or it falls back to just wrapping the object itself like any secure coding type.

That NSDictionary is known to contain NSValues that wrap an NSRange, so this patch also adds an NSValue wrapper.

Currently, any NSValues that are serialized would go through the secure coding path.
For this one wrapped type we're known to need (NSRange), we go through a directly serialized code path.

Headed down this road, the goal is to identify all specific types we need to serialize, have a direct wrapper
for each, and reduce our reliance on the secure coding code path to zero.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.h:
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::getClass<DDScannerResult>):
(IPC::typeFromObject):
* Source/WebKit/Shared/Cocoa/CoreIPCDDScannerResult.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCDictionary.h.
(WebKit::CoreIPCDDScannerResult::CoreIPCDDScannerResult):
* Source/WebKit/Shared/Cocoa/CoreIPCDDScannerResult.mm: Copied from Source/WebKit/Shared/Cocoa/CoreIPCSecureCoding.mm.
(WebKit::shouldWrapDDScannerResult):
(WebKit::CoreIPCDDScannerResult::CoreIPCDDScannerResult):
(WebKit::CoreIPCDDScannerResult::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCDDScannerResult.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCDictionary.h:
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.h:
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.mm:
(WebKit::valueFromID):
* Source/WebKit/Shared/Cocoa/CoreIPCNSValue.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCDictionary.h.
(WebKit::CoreIPCNSValue::CoreIPCNSValue):
* Source/WebKit/Shared/Cocoa/CoreIPCNSValue.mm: Copied from Source/WebKit/Shared/Cocoa/CoreIPCSecureCoding.mm.
(WebKit::CoreIPCNSValue::CoreIPCNSValue):
(WebKit::CoreIPCNSValue::toID const):
(WebKit::CoreIPCNSValue::shouldWrapValue):
* Source/WebKit/Shared/Cocoa/CoreIPCNSValue.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCSecureCoding.h:
(WebKit::CoreIPCSecureCoding::CoreIPCSecureCoding):
* Source/WebKit/Shared/Cocoa/CoreIPCSecureCoding.mm:
(WebKit::CoreIPCSecureCoding::CoreIPCSecureCoding):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.mm:
(IPC::getClass<DDScannerResult>): Deleted.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Since NSNumber is itself an NSValue, refactor these tests to explicitly emplace into the testing variant:
* Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
(TEST):
(fakeDataDetectorResultForTesting):

Canonical link: https://commits.webkit.org/270840@main
